controllers/release/release_adapter_test.go

```diff
@@ -878,25 +842,121 @@ var _ = Describe("Release Adapter", Ordered, func() {
 
-		// Restore original targetContext value
-		adapter.targetContext = adapter.context
+			// Add a target workspace to the release Status
+			release.Status.Target = kcp.NamespaceReference{
+				Namespace: "default",
+				Workspace: "test-workspace",
+			}
+			Expect(k8sClient.Status().Patch(ctx, release, patch)).To(Succeed())
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, types.NamespacedName{
+					Name:      release.Name,
+					Namespace: release.Namespace,
+				}, release)
+				return err == nil && release.Status.Target.Workspace == "test-workspace"
+			}, time.Second*10).Should(BeTrue())
+
+			result, err := adapter.EnsureTargetContextIsSet()
+			Expect(err).NotTo(HaveOccurred())
+			Expect(!result.RequeueRequest).To(BeTrue())
+
+			clusterName, _ := logicalcluster.ClusterFromContext(adapter.targetContext)
+			Expect(clusterName.String()).To(Equal("test-workspace"))
+
+			// Restore original values
+			adapter.targetContext = adapter.context
+		})
+
+		It("should mark release as invalid when the target workspace is set on a non-KCP environment", func() {
+			// Add a target workspace to the ReleasePlan
+			patch := client.MergeFrom(releasePlan.DeepCopy())
+			originalWorkspace := releasePlan.Spec.Target.Workspace
+			releasePlan.Spec.Target.Workspace = "test-workspace"
+			Expect(k8sClient.Patch(ctx, releasePlan, patch)).Should(Succeed())
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, types.NamespacedName{
+					Name:      releasePlan.Name,
+					Namespace: releasePlan.Namespace,
+				}, releasePlan)
+				return err == nil && releasePlan.Spec.Target.Workspace == "test-workspace"
+			}, time.Second*10).Should(BeTrue())
+
+			result, err := adapter.EnsureTargetContextIsSet()
+			Expect(result.CancelRequest).To(BeTrue())
+			Expect(err).NotTo(HaveOccurred())
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, types.NamespacedName{
+					Name:      release.Name,
+					Namespace: release.Namespace,
+				}, release)
+				return err == nil && len(release.Status.Conditions) > 0
+			}, time.Second*10).Should(BeTrue())
+
+			Expect(release.Status.Conditions[0].Reason).To(Equal("ReleasePlanValidationError"))
+
+			// Restore original values
+			adapter.targetContext = adapter.context
+			patch = client.MergeFrom(releasePlan.DeepCopy())
+			releasePlan.Spec.Target.Workspace = originalWorkspace
+			Expect(k8sClient.Patch(ctx, releasePlan, patch)).Should(Succeed())
+		})
+
+		It("ensures the target context is set to default context when releaseplan does not contain a target workspace", func() {
+			result, err := adapter.EnsureTargetContextIsSet()
+			Expect(!result.RequeueRequest && err == nil).To(BeTrue())
+			Expect(adapter.targetContext).To(Equal(adapter.context))
+		})
+
+		It("ensure an error is returned when setting the target context with a missing releaseplan", func() {
+			// Delete the ReleasePlan
+			Expect(k8sClient.Delete(ctx, releasePlan)).Should(Succeed())
```

Comment:
This is why I wanted to have the releasePlan (and admission) created in BeforeEach instead of BeforeAll. This test *must* remain as our last test in the file now

Reply:
Can we at least have `Expect(k8sClient.Create(ctx, releasePlan)).Should(Succeed())` after line 958 for now then?

Reply:
Sure. Added.